src/include/mallocMC/mallocMC_utils.hpp

```diff
@@ -186,4 +186,39 @@ namespace mallocMC
   template<class T>
   MAMC_HOST MAMC_ACCELERATOR inline T divup(T a, T b) { return (a + b - 1)/b; }
 
+  /** the maximal number threads per block
+   *
+   * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities
+   */
+  struct MaxThreadsPerBlock
+  {
+	// valid for sm_2.X - sm_7.5
+    BOOST_STATIC_ASSERT uint32_t value = 1024;
+  };
+
+  /** the maximal number threads per block
+   *
+   * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities
+   */
+  struct WarpSize
+  {
+	// valid for sm_2.X - sm_7.5
+    BOOST_STATIC_ASSERT uint32_t value = 32;
+  };
+
+  /** warp id within a block
+   *
+   * The id is constant over the livetime of the thread.
+   * The id is not equal to warpid().
+   *
+   * @return warp id within the block
+   */
+  MAMC_ACCELERATOR inline boost::uint32_t warpid_withinblock()
+  {
+    return (
+      threadIdx.z * blockDim.y * blockDim.x +
```

Comment:
that implementation is likely not true if any of he block dims is not exactly a multiple of the warp size.

Reply:
```
For this reason, %ctaid and %tid should be used to compute a virtual warp index if such a value is needed in kernel code
```

Reply:
btw, didn't know that `%warpid` is also quite expensive (besides being wrong for our case): https://devtalk.nvidia.com/default/topic/1011523/cuda-programming-and-performance/how-costly-is-the-s2r-instruction-reading-a-special-register-/post/5165296/#5165296

Reply:
> that implementation is likely not true if any of the block dims is not exactly a multiple of the warp size.

It is not required that it is a multiple. It gives the linear thread idx and than device it by the warp size.
Could you please explain why it should be wrong?

Reply:
offline discussion: the question is just about the strategy of thread-linearization to warps at runtime. do we have a reference that this is the same mapping during thread scheduling?